### PR TITLE
Fix growpart race

### PR DESCRIPTION
**Pull request by holmanb**
## Proposed Commit Message
Fixes https://github.com/canonical/cloud-init/issues/4613

## Additional Context
See bug

## Checklist

- [ ] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/development/index.html)
- [ ] I have updated or added any [unit tests](https://cloudinit.readthedocs.io/en/latest/development/testing.html) accordingly
- [ ] I have updated or added any [documentation](https://cloudinit.readthedocs.io/en/latest/development/contribute_docs.html) accordingly

## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

### CODE REVIEW

File: cloudinit/config/cc_growpart.py [hunk 1]
```diff
@@ -284,11 +284,15 @@ def resize(self, diskdev, partnum, partdev):
 
 
 def get_size(filename):
```
Comment (TheRealFalcon) [inline cloudinit/config/cc_growpart.py:286]: Let's add an `Optional[int]` return type here since we're changing the return type

**Review by TheRealFalcon — COMMENTED** [review]
2 minor comments inline.

Can we make sure this is unit tested?

File: cloudinit/config/cc_growpart.py [hunk 2]
```diff
@@ -580,6 +584,15 @@ def resize_devices(resizer, devices):
                         "no change necessary (%s, %s)" % (disk, ptnum),
                     )
                 )
+            elif new is None or old is None:
+                info.append(
+                    (
+                        devent,
+                        RESIZE.CHANGED,
+                        "changed (%s, %s) size, new size is unknown"
+                        % (disk, ptnum, old),
```
Comment (TheRealFalcon) [inline cloudinit/config/cc_growpart.py:593]: Now there's too many arguments here

**Comment by holmanb** [comment]
Thanks @TheRealFalcon. This one fell off my radar - in part due to test complexity.

`resize_devices()` will require lots of mocks in order to get coverage for the log message logic - enough that I don't think that such a heavily mocked test provides much value. Refactoring that function would likely be required to make it more reasonable to write tests for, but I'm hoping to avoid that for now.

I've pushed a test for `get_size()`, let me know if that suffices for now.

**Comment by TheRealFalcon** [comment]
Yeah, that one's a bit of a doozy. Given that we have some integration test coverage (that triggered this issue), I'm ok skipping that function for now.

**Review by TheRealFalcon — APPROVED** [review]

**Pull request merged**